### PR TITLE
Checksums calculated with full file checksums

### LINKED ISSUE #17920: Read checksum of SST files from RocksDB instead of computing it
**Description**

When taking a new snapshot from local runtime, Zeebe will create a SFV file containing the checksum of each file. To get it, it actually reads each file synchronously, meaning the time it takes to create this file is proportional to the size of the snapshot. If you have large snapshots, this can be quite expensive.

Thankfully, the bulk of the snapshot files are the SST files, and RocksDB now offers a new API to get the pre-computed checksum directly without having to read the files.

To do this, you need to:

1. Take the snapshot
2. Open the snapshot as a read-only DB
3. Request the snapshot for the SST files (`GetLiveFilesMetadata`)
4. Read and compute snapshot for non-SST files

As the bulk of the snapshot is SST files, reading the other files is not proportional to the size of the DB, but is bounded by a few configuration options (e.g. max manifest size and so on).

As we will be backporting this all the way to 8.3, we will also likely have to backport the RocksDB version, as this API is fairly new.

**Context**

https://jira.camunda.com/browse/SUPPORT-21546

### PR DESCRIPTION
**Pull request by EuroLew**
## Description

Add full file checksum functionality.

Currently will not past tests due to the combined checksum verification still occurring, and full file checksum not being enabled. These issues are fixed in the #18722 and #18721 

part 3/3/ to address https://github.com/camunda/zeebe/issues/17920
## Related issues

closes #17920

### CODE REVIEW

File: zeebe/restore/src/main/java/io/camunda/zeebe/restore/PartitionRestoreService.java
Comment (oleschoenburg) [inline zeebe/restore/src/main/java/io/camunda/zeebe/restore/PartitionRestoreService.java:61]: :wrench: These constructors are a bit confusing. I'd suggest to always have one "full" constructor which sets all the fields and then convenience constructors which all delegate to the full constructor and assign default values to some parameters.

File: zeebe/restore/src/main/java/io/camunda/zeebe/restore/PartitionRestoreService.java
Comment (oleschoenburg) [inline zeebe/restore/src/main/java/io/camunda/zeebe/restore/PartitionRestoreService.java:193]: :wrench: Let's do this fallback to `ChecksumProviderRocksDBImpl` in the constructors.

File: zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/MutableChecksumsSFV.java
Comment (oleschoenburg) [inline zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/MutableChecksumsSFV.java:49]: :upside_down_face: 

```suggestion
  void updateFromChecksum(final Path filePath, final long checksum);
```

Always a bit nicer to use the primitive types instead of the boxed ones.

File: zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
Comment (oleschoenburg) [inline zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java:81]: :x: this should be final.

File: zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
Comment (oleschoenburg) [inline zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java:90]: :wrench: Same thing with the constructors here. It's easier to understand if there's one full constructor that all others delegate to.

File: zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/SnapshotChecksum.java
Comment (oleschoenburg) [inline zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/SnapshotChecksum.java:52]: :wrench: I think the name doesn't quite fit. It's more like `calculateWithProvidedChecksums`, right?

File: zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/SnapshotChecksum.java
Comment (oleschoenburg) [inline zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/SnapshotChecksum.java:106]: :upside_down_face: Either use `var` or the primitive `int` here.
:wrench: I'd move the `Integer.toUnsignedLong` to this line too. Otherwise it's not the "real" `sstChecksum`.

File: zeebe/zb-db/src/main/java/io/camunda/zeebe/db/ChecksumProvider.java
Comment (oleschoenburg) [inline zeebe/zb-db/src/main/java/io/camunda/zeebe/db/ChecksumProvider.java:20]: :x: I think it would be better to have this return `Map<String, Long>`. The conversion from big-endian int to unsigned long is specific to RocksDb and should be encapsulated in the implementation and not leak into the interface.

Comment (oleschoenburg) [inline zeebe/zb-db/src/main/java/io/camunda/zeebe/db/ChecksumProvider.java:20]: :upside_down_face: The `public` is redundant, IntelliJ should also highlight that.

File: zeebe/restore/pom.xml
Comment (oleschoenburg) [inline zeebe/restore/pom.xml:33]: :thought_balloon: Instead of adding a dependency on `zeebe-db` to make `ChecksumProviderRocksDBImpl available, we could also promote the `TestChecksumProvider` to a `NoopChecksumProvider` and use that instead. We don't actually need the provider during restore so I think I'd favor this.

**Review by lenaschoenburg — CHANGES_REQUESTED** [review]
Thanks, this is overall a really neat PR :+1: 

I have some suggestions and requested changes, please take a look. Nothing major though.

File: zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/ChecksumProvider.java
Comment (oleschoenburg) [inline zeebe/snapshot/src/main/java/io/camunda/zeebe/snapshots/ChecksumProvider.java:18]: :wrench: The documentation is out of date now

**Comment by EuroLew** [comment]
Don't actually need restore to have any knowledge of checksum provider. Since it doesn't use the RocksDBImpl and just the Noop implementation just leaving the code as it is an instantiating a null providing as the default will result in the old behaviour 
image

**Review by lenaschoenburg — APPROVED** [review]
Nice work :+1: 

:wrench: Could you add a test that shows that `ChecksumProvider` is actually used? My idea on how to test this is to create a file with random content and a `ChecksumProvider` that gives an obviously wrong checksum for that file. Then assert that the SfV checksum contains the "wrong" checksum and not the one calculated over the file contents.

**Comment by EuroLew** [comment]
> Nice work 👍
> 
> 🔧 Could you add a test that shows that `ChecksumProvider` is actually used? My idea on how to test this is to create a file with random content and a `ChecksumProvider` that gives an obviously wrong checksum for that file. Then assert that the SfV checksum contains the "wrong" checksum and not the one calculated over the file contents.

Added `shouldUseChecksumProviderForChecksumsIfSupplied` test.

File: zeebe/snapshot/src/test/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStoreTest.java
Comment (oleschoenburg) [inline zeebe/snapshot/src/test/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStoreTest.java:262]: :upside_down_face: please split this up into given, when and then. You can take a look at the other tests for examples.

**Review by lenaschoenburg — APPROVED** [review]
Thanks :+1: 

In general, once you got an approval you can feel free to make minor changes (such as adding a test, renaming things etc.) without another review cycle. It's up to you to decide whether you _want_ another review or not.

**Comment by EuroLew** [comment]
8.5 backport seems to be working 
image
image

File: zeebe/zb-db/src/main/java/io/camunda/zeebe/db/ChecksumProvider.java
Comment (npepinpe) [reply zeebe/zb-db/src/main/java/io/camunda/zeebe/db/ChecksumProvider.java]: Just stumbled on this, but in the future, we can also use `Object2LongHashMap` from Agrona to also not deal with boxed types ;)

**Pull request merged**
